api-proposal.md

```diff
@@ -1,90 +1,152 @@
-# Proposal for ecosystem website api.
+# Proposal for Ecosystem Website API
 
 ## API's
 
 ---
 
-### Package List
+### Error handling
 
-We could have some specifc string matching for categories and tags in the query if thats what the user is searching for to maybe speed things up. `cat:<category>` and `tag:<string>`.
+On error, each endpoint returns a HTTP status code 500, and the following body
+```javascript
+data: {
+  error: string // containing an error message to be displayed in the UI
+}
+```
 
-- GET `/api/v1/packages?query=:string&page=:number`
-  ```
-  items: [<package>]
-  ```
+### Authentication
 
----
+An authenticated user is authenticated against GitHub.
+An authenticated administrator is defined by a set of GitHub usernames in the configuration of the backend.
+
+### Endpoints
 
-### Package
+#### Package
 
 - **Schema**
 
-  ```
+  ```javascript
   <package> {
-    id: number
-    name: string
-    description: string
-    readme: string // maybe from github or something?
-    image: blob || string
+    id: string // a unique, immutable string identifiying this package. allowed: [a-z-_]{2,}
+    name: string // human readable name
+    description: string 
+    readme: string // user can enter a markdown description
+    image: blob // we should restrict the image size to 5kb
     website: string
     repository: string
     license: string
     commentsCount: number
     upvotes: number
     downvotes: number
-    tags: string[]
+    tags: string[] // tags are created automatically if they don't exist yet. Only [a-z-_]{2,} is allowed for tags.
     added: datetime
     updated: datetime
+    owner: string // github username
   }
 
   ```
 
-- GET `/api/v1/packages/:package`
-  ```
+- **Get a package by id**
+
+  GET `/api/v1/packages/:package`
+  ```javascript
   data: {
     package: <package>
-    comments: [<comment>] // first 50(?) comments
-    releases: [<release>] // first 10(?) releases
   }
   ```
-- POST `/api/v1/packages`
+  
+  
+- **Get packages a by search query**
+  
+  GET `/api/v1/packages?query=:queryString&page=:number&order=:order&direction=:dir`
+  ```javascript
+  data: {
+    package: [<package>] // array of packages matching the queryString
+    resultCount: <number> // total number of search query results
+  }
+  ```
+  The `:queryString` supports a special syntax for selecting packages based on a tag: `tag:<string>`. No (or an empty query string) returns all packages.
```

Comment:
Markdown didn't preserve your line breaks here, I think you can add 2 blank spaces to the end, or just make this a second level bullet list.